sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -752,61 +744,79 @@ public void processElement(ProcessContext context) throws IOException {
-                            }
-                            LOG.debug("Done finalizing write operation for {}.", entry.getKey());
+                  ParDo.of(new DoFn<FileResult<DestinationT>, KV<DestinationT, String>>() {
+                      private List<FileResult<DestinationT>> fileResults;
+                      @Nullable private Integer fixedNumShards;
+
+                      @StartBundle
+                      public void startBundle() {
+                        fileResults = Lists.newArrayList();
+                        fixedNumShards = null;
+                      }
+
+                      @ProcessElement
+                      public void processElement(ProcessContext c) {
+                        fileResults.add(c.element());
+                        if (fixedNumShards == null && fixedNumShardsView != null) {
+                          fixedNumShards = c.sideInput(fixedNumShardsView);
+                        }
+                      }
+
+                      @FinishBundle
+                      public void finishBundle(FinishBundleContext c) throws Exception {
+                        Set<ResourceId> tempFiles = Sets.newHashSet();
+                        Multimap<DestinationT, FileResult<DestinationT>> results =
+                            perDestinationResults(fileResults);
+                        for (Map.Entry<DestinationT, Collection<FileResult<DestinationT>>>
+                            destEntry : results.asMap().entrySet()) {
+                          LOG.info(
+                              "Finalizing write operation {} for destination {} num shards: {}.",
+                              writeOperation,
+                              destEntry.getKey(),
+                              destEntry.getValue().size());
+                          List<KV<FileResult<DestinationT>, ResourceId>> resultsToFinalFilenames =
+                              writeOperation.finalize(fixedNumShards, destEntry.getValue());
+                          for (KV<FileResult<DestinationT>, ResourceId> entry
+                              : resultsToFinalFilenames) {
+                            FileResult<DestinationT> res = entry.getKey();
+                            BoundedWindow window = res.getWindow();
+                            c.output(
+                                KV.of(res.getDestination(), entry.getValue().toString()),
+                                window.maxTimestamp(),
+                                window);
                           }
-                          writeOperation.removeTemporaryFiles(tempFiles);
-                          LOG.debug("Removed temporary files for {}.", writeOperation);
+                          LOG.debug(
+                              "Done finalizing write operation for {}.", destEntry.getKey());
                         }
-                      }))
+                        writeOperation.removeTemporaryFiles(tempFiles);
```

Comment:
I think you dropped the code that adds elements to tempFiles :)

Reply:
Oops. Added a test.